fluent-bundle/src/builtins.ts

```diff
@@ -0,0 +1,67 @@
+/**
+ * @overview
+ *
+ * The FTL resolver ships with a number of functions built-in.
+ *
+ * Each function take two arguments:
+ *   - args - an array of positional args
+ *   - opts - an object of key-value args
+ *
+ * Arguments to functions are guaranteed to already be instances of
+ * `FluentType`.  Functions must return `FluentType` objects as well.
+ */
+
+import {
+  FluentNone,
+  FluentNumber,
+  FluentDateTime,
+  FluentType
+} from "./types.js";
+
+function values(opts: Record<string, FluentType>) {
+  const unwrapped: Record<string, unknown> = {};
+  for (const [name, opt] of Object.entries(opts)) {
+    unwrapped[name] = opt.valueOf();
+  }
+  return unwrapped;
+}
+
+export function NUMBER(
+  args: Array<FluentType>,
+  opts: Record<string, FluentType>
+) {
+  if (args.length !== 1) {
+    throw new TypeError("NUMBER accepts exactly one argument");
+  }
+  let arg = args[0];
+
+  if (arg instanceof FluentNone) {
+    return new FluentNone(`NUMBER(${arg.valueOf()})`);
+  }
+
+  if (arg instanceof FluentNumber || arg instanceof FluentDateTime) {
+    return new FluentNumber(arg.valueOf(), { ...arg.opts, ...values(opts) });
+  }
+
+  throw new TypeError("Invalid argument to NUMBER");
```

Comment:
Unfortunately, the amount of changes in this file makes diff think it's a new file. Here's what `NUMBER` used to look like:

```js
export function NUMBER([arg], opts) {
  if (arg instanceof FluentNone) {
    return new FluentNone(`NUMBER(${arg.valueOf()})`);
  }

  let value = Number(arg.valueOf());
  if (Number.isNaN(value)) {
    throw new TypeError("Invalid argument to NUMBER");
  }

  return new FluentNumber(value, { ...arg.opts, ...values(opts) });
}
```

This code didn't compile in TS because `arg.opts` doesn't exist on `FluentBaseType`; it's only defined on `FluentNumber` and `FluentDateTime`. BTW. the JS code works because `{...undefined}` evaluates to `{}`.

In the TS rewrite, I introduced explicit `arg instanceof FluentNumber || arg instanceof FluentDateTime` guards and dropped the `Number()` parsing. This changes the current behavior of `@fluent/bundle`: calling `NUMBER` on a string-typed variable used to parse the string as a number and format it. After this change, the argument will be rejected as invalid.

Reply:
I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1605062 to prepare for this change in `mozilla-central`.